launch/launch/utilities/type_utils.py

```diff
@@ -0,0 +1,567 @@
+
+"""Tuple of valid scalar types."""
+ScalarTypesTuple = (
+    int, float, bool, str
+)
+
+"""Tuple of valid list types."""
+ListTypesTuple = (
+    List[int], List[float], List[bool], List[str]
+)
+
+AllowedTypesTuple = ScalarTypesTuple + ListTypesTuple
+
+"""Allowed scalar types."""
+ScalarTypesType = Type[Union[int, float, bool, str]]
+"""Allowed uniform list types."""
+ListTypesType = Type[Union[
+    List[int], List[float], List[bool], List[str]
+]]
+"""Allowed types."""
+AllowedTypesType = Union[ScalarTypesType, ListTypesType]
+
+"""Allowed scalar values."""
+ScalarValueType = Union[int, float, bool, str]
+"""Allowed uniform list values."""
+ListValueType = Union[List[int], List[float], List[bool], List[str]]
+"""Allowed uniform sequence values."""
+SequenceValueType = Union[Sequence[int], Sequence[float], Sequence[bool], Sequence[str]]
+"""Allowed values."""
+AllowedValueType = Union[ScalarValueType, SequenceValueType]
+
+"""Substitution to a scalar type or a scalar type."""
+SomeScalarType = Union[SomeSubstitutionsType, ScalarValueType]
+"""Substitution that can be performed to a uniform list."""
+SomeSequenceType = Union[(
+    Sequence[Union[int, SomeSubstitutionsType]],
+    Sequence[Union[float, SomeSubstitutionsType]],
+    Sequence[Union[bool, SomeSubstitutionsType]],
+    Sequence[Union[str, SomeSubstitutionsType]],
+)]
+"""Union of SomeScalarType and SomeSequenceType."""
+SomeValueType = Union[SomeScalarType, SomeSequenceType]
+
+"""Normalized version of SomeScalarType."""
+NormalizedScalarType = Union[List[Substitution], ScalarValueType]
+"""Normalized version of SomeSequenceType."""
+NormalizedSequenceType = Union[
+    List[Union[int, List[Substitution]]],
+    List[Union[float, List[Substitution]]],
+    List[Union[bool, List[Substitution]]],
+    List[Union[str, List[Substitution]]],
+]
+"""Normalized version of SomeValueType."""
+NormalizedValueType = Union[NormalizedScalarType, NormalizedSequenceType]
+
+"""String embedded substitution version of SomeScalarType"""
+StrSomeScalarType = SomeScalarType
+"""String embedded substitution version of SomeSequenceType"""
+StrSomeSequenceType = Union[(
+    Sequence[Union[int, str]],
+    Sequence[Union[float, str]],
+    Sequence[Union[bool, str]],
+    Sequence[Union[str]],
+)]
+"""String embedded substitution version of SomeValueType"""
+StrSomeValueType = Union[StrSomeScalarType, StrSomeSequenceType]
+
+
+def is_typing_list(data_type: Any) -> bool:
+    """
+    Return `True` if data_type is `typing.List` or a subscription of it.
+
+    Examples
+    --------
+    ```python3
+    assert is_typing_list(typing.List)
+    assert is_typing_list(typing.List[int])
+    assert not is_typing_list(int)
+    ```
+
+    """
+    return data_type is List or (
+        hasattr(data_type, '__origin__') and
+        hasattr(data_type, '__args__') and
+        data_type.__origin__ in  # type: ignore
+        (list, List) and  # On Linux/Mac is List, on Windows is list.
```

Comment:
Note to myself: this is crazy. I wonder why.

Reply:
This code was already there before this PR.
I think it wasn't platform dependent, but we where using at that time a different Python version on Windows.

The Python `typing` module is super new and still being developed, so it's not weird that this kind of detail changed.
I guess that the latest version is using `List`, and not `list`.

Reply:
> The Python typing module is super new and still being developed, so it's not weird that this kind of detail changed.

Yeah, that I know. To have a type change on a platform-basis is, nonetheless, curious. Looking at the `typing` module sources, it looks pretty platform-independent to me.

Reply:
> Yeah, that I know. To have a type change on a platform-basis is, nonetheless, curious. Looking at the typing module sources, it looks pretty platform-independent to me.

Yes, it's platform independent, but it changed between Python versions.
The comment is misleading, I will correct it.

Reply:
Done in https://github.com/ros2/launch/pull/438/commits/9ab3835a7d4f37b212858fd0744ca409bc424cb0

Reply:
> Yes, it's platform independent, but it changed between Python versions.

Aha! Now it makes sense.